### ISSUE TITLE
Integration with Vimspector for debugging

### ISSUE THREAD

**Issue by wmvanvliet** [jupyter-vim/jupyter-vim#17]
Debugging in Jupyter is still a bit awkward, since readline keybindings are not supported when talking to pdb.

Maybe we should consider adding integration with [Vimspector](https://github.com/puremourning/vimspector) for a better debugging experience.

**Comment by tinmarino** [comment]
What is your idea ?

1. Vim -> Jupyter: sending the debugged environment to jupyter to that you can evaluate the current scope in console
2. Jupyter -> Vim: sending some next commands from jupyter and update vimspector state

Is that it ?

**Comment by wmvanvliet** [comment]
Seems like the new Jupyter kernel has dedicated commands for debugging. We should probably look into that.

**Comment by wmvanvliet** [comment]
The new release now has a debugger protocol! This should make it possible for us to integrate debugging. https://discourse.jupyter.org/t/ann-ipykernel-6-0-release/9804

**Comment by wmvanvliet** [comment]
> What is your idea ?
> 
>     1. Vim -> Jupyter: sending the debugged environment to jupyter to that you can evaluate the current scope in console
> 
>     2. Jupyter -> Vim: sending some next commands from jupyter and update vimspector state
> 
> 
> Is that it ?

Yes. That is essentially it. I've been playing around a bit with it and it is in theory possible to hook them up together. Here is a screencast that shows this off:

https://www.youtube.com/watch?v=7GtEbELYeUE

With the newest version of ipykernel, a debugpy debugger is started alongside the python kernel. The jupyter messaging protocol allows access to this debugger using the DAP protocol, wrapped inside jupyter messages.

My first try was to connect vimspector to the debugpy instance directly, but apparently it only allows a single connection (?), which is taken up by the ipython kernel. My second try was writing a small script that acts as DAP server, forwarding all DAP messages to the ipython kernel using the jupyter_client package. Vimspector can then be pointed at this script, yielding the result you see in the screencast.

Ideally, I'm striving for a tight coupling between the console and vimspector. If you type `%debug` in the console, vimspector launches in your vim instance. If you hit the debug button in vim, the console switches to debug mode, sending all commands to the debugpy REPL. When you select some lines in vim and hit `:JupyterDebug`, these lines should start to execute and whenever a breakpoint is hit, vimspector should open at that breakpoint. There's a long way to go here. Ultimately I want the debugging experience in vim to be as great as in MATLAB.
